File: WaveTheory.java
/*     */ import java.applet.Applet;
/*     */ import java.awt.BorderLayout;
/*     */ import java.awt.Button;
/*     */ import java.awt.Choice;
/*     */ import java.awt.Event;
/*     */ import java.awt.GridLayout;
/*     */ import java.awt.Label;
/*     */ import java.awt.Panel;
/*     */ import java.awt.TextField;
/*     */ 
/*     */ 
/*     */ 
/*     */ public class WaveTheory
/*     */   extends Applet
/*     */ {
/*     */   Panel panel1;
/*     */   Panel results;
/*     */   Label label1;
/*     */   TextField Hin;
/*     */   TextField Tin;
/*     */   TextField din;
/*     */   TextField angle0;
/*     */   Button Calculate;
/*     */   Button Reset;
/*     */   Refract ref;
/*     */   Choice torf;
/*     */   int ntheory;
/*     */   
/*     */   public void init() {
/*  30 */     resize(450, 450);
/*  31 */     setLayout(new BorderLayout());
/*  32 */     this.panel1 = new Panel();
/*  33 */     this.panel1.setLayout(new GridLayout(9, 2));
/*  34 */     this.panel1.add(new Label("Deep Water Values:"));
/*  35 */     this.panel1.add(new Label(" "));
/*  36 */     this.panel1.add(new Label("Wave Height (m)?"));
/*  37 */     this.panel1.add(this.Hin = new TextField(""));
/*  38 */     this.torf = new Choice();
/*  39 */     this.torf.addItem("Period");
/*  40 */     this.torf.addItem("Frequency (Hz)");
/*  41 */     this.panel1.add(this.torf);
/*  42 */     this.panel1.add(this.Tin = new TextField(""));
/*  43 */     this.panel1.add(new Label("Wave Angle (o)?"));
/*  44 */     this.panel1.add(this.angle0 = new TextField("0.0"));
/*  45 */     this.panel1.add(new Label("Local Depth?"));
/*  46 */     this.panel1.add(this.din = new TextField(""));
/*  47 */     this.panel1.add(new Label(""));
/*  48 */     this.panel1.add(new Label(""));
/*  49 */     this.panel1.add(new Label(""));
/*  50 */     this.panel1.add(new Label(""));
/*  51 */     this.panel1.add(this.Calculate = new Button("Calculate"));
/*  52 */     this.panel1.add(this.Reset = new Button("Reset"));
/*  53 */     this.panel1.add(new Label(""));
/*  54 */     this.panel1.add(new Label(""));
/*  55 */     add("West", this.panel1);
/*  56 */     this.results = new Panel();
/*  57 */     this.results.setLayout(new GridLayout(10, 2));
/*  58 */     this.results.add(new Label(" L (m) ="));
/*  59 */     this.lout = new TextField("           ");
/*  60 */     this.lout.setEditable(false);
/*  61 */     this.results.add(this.lout);
/*  62 */     this.results.add(new Label(" k=2 pi/L = "));
/*  63 */     this.kout = new TextField("           ");
/*  64 */     this.kout.setEditable(false);
/*  65 */     this.results.add(this.kout);
/*  66 */     this.results.add(new Label(" C =L/T = "));
/*  67 */     this.Cout = new TextField("           ");
/*  68 */     this.Cout.setEditable(false);
/*  69 */     this.results.add(this.Cout);
/*  70 */     this.results.add(new Label("  Cg  ="));
/*  71 */     this.Cgout = new TextField("           ");
/*  72 */     this.Cgout.setEditable(false);
/*  73 */     this.results.add(this.Cgout);
/*  74 */     this.results.add(new Label("  n= Cg/C = "));
/*  75 */     this.nout = new TextField("            ");
/*  76 */     this.nout.setEditable(false);
/*  77 */     this.results.add(this.nout);
/*  78 */     this.results.add(new Label("  Ks = "));
/*  79 */     this.Ksout = new TextField("            ");
/*  80 */     this.Ksout.setEditable(false);
/*  81 */     this.results.add(this.Ksout);
/*  82 */     this.results.add(new Label("  Kr = "));
/*  83 */     this.Krout = new TextField("              ");
/*  84 */     this.Krout.setEditable(false);
/*  85 */     this.results.add(this.Krout);
/*  86 */     this.results.add(new Label(" Angle = "));
/*  87 */     this.angleout = new TextField("           ");
/*  88 */     this.angleout.setEditable(false);
/*  89 */     this.results.add(this.angleout);
/*  90 */     this.results.add(new Label("   H = "));
/*  91 */     this.Hshal = new TextField("          ");
/*  92 */     this.Hshal.setEditable(false);
/*  93 */     this.results.add(this.Hshal);
/*  94 */     this.results.add(new Label("  u_b= "));
/*  95 */     this.ubout = new TextField("        ");
/*  96 */     this.ubout.setEditable(false);
/*  97 */     this.results.add(this.ubout);
/*  98 */     add("Center", this.results);
/*     */   }
/*     */ 
/*     */   
/*     */   TextField lout;
/*     */   TextField Cgout;
/*     */   TextField Ksout;
/*     */   TextField Krout;
/*     */   TextField angleout;
/*     */   TextField fin;
/*     */   
/*     */   public boolean handleEvent(Event paramEvent) {
/* 110 */     this.ref = new Refract();
/*     */     
/*     */     try {
/* 113 */       if (paramEvent.id == 1001) {
/* 114 */         if (paramEvent.target.equals(this.torf)) {
/* 115 */           if ("Period".equals(paramEvent.arg)) {
/* 116 */             this.period = true;
/*     */           } else {
/*     */             
/* 119 */             this.period = false;
/*     */           } 
/*     */         }
/*     */         
/* 123 */         if (paramEvent.target instanceof Button) {
/* 124 */           if (paramEvent.target == this.Calculate) {
/* 125 */             float f2, f1 = Float.valueOf(this.Hin.getText()).floatValue();
/* 126 */             String str = this.Tin.getText();
/* 127 */             if (this.period) {
/* 128 */               f2 = Float.valueOf(str).floatValue();
/*     */             } else {
/*     */               
/* 131 */               f2 = (float)(1.0D / Float.valueOf(str).floatValue());
/*     */             } 
/* 133 */             float f3 = Float.valueOf(this.din.getText()).floatValue();
/* 134 */             float f12 = Float.valueOf(this.angle0.getText()).floatValue();
/* 135 */             float f5 = this.ref.waveNumber(f3, f2);
/* 136 */             this.kout.setText(String.valueOf(f5));
/* 137 */             float f4 = 6.283185F / f5;
/* 138 */             this.lout.setText(String.valueOf(f4));
/* 139 */             float f10 = f4 / f2;
/* 140 */             this.Cout.setText(String.valueOf(f10));
/* 141 */             float f6 = this.ref.theta(f12, f5, f2);
/* 142 */             this.angleout.setText(String.valueOf(f6));
/* 143 */             float f7 = this.ref.groupVelocity(f2, f3, f5);
/* 144 */             float f11 = f7 / f10;
/* 145 */             this.nout.setText(String.valueOf(f11));
/* 146 */             this.Cgout.setText(String.valueOf(f7));
/* 147 */             float f8 = this.ref.shoalingCoef(f2, f7);
/* 148 */             this.Ksout.setText(String.valueOf(f8));
/* 149 */             float f9 = this.ref.refractionCoef(f12, f6);
/* 150 */             this.Krout.setText(String.valueOf(f9));
/* 151 */             float f13 = f8 * f9 * f1;
/* 152 */             if (f13 > 0.8F * f3) {
/* 153 */               f13 = 0.8F * f3;
/* 154 */               this.Hshal.setText(String.valueOf(String.valueOf(f13)) + ", breaking");
/*     */             }
/*     */             else {
/*     */               
/* 158 */               this.Hshal.setText(String.valueOf(f13));
/*     */             } 
/* 160 */             double d = 6.283185307179586D / f2 * f13 / 2.0D * Hyper.sinh((f5 * f3));
/* 161 */             this.ubout.setText(String.valueOf(d));
/*     */           
/*     */           }
/* 164 */           else if (paramEvent.target == this.Reset) {
/* 165 */             this.Hin.setText("");
/* 166 */             this.Tin.setText("");
/* 167 */             this.din.setText("");
/* 168 */             this.angle0.setText("");
/* 169 */             this.lout.setText("");
/* 170 */             this.kout.setText("");
/* 171 */             this.Cout.setText("");
/* 172 */             this.Cgout.setText("");
/* 173 */             this.nout.setText("");
/* 174 */             this.Ksout.setText("");
/* 175 */             this.Krout.setText("");
/* 176 */             this.kout.setText("");
/* 177 */             this.angleout.setText("");
/* 178 */             this.Hshal.setText("");
/* 179 */             this.ubout.setText("");
/*     */           }
/*     */         
/*     */         }
/*     */       }
/*     */     
/* 185 */     } catch (NumberFormatException numberFormatException) {
/*     */       
/* 187 */       System.out.println(numberFormatException);
/*     */     } 
/*     */     
/* 190 */     return false;
/*     */   }
/*     */   
/*     */   TextField Hshal;
/*     */   TextField nout;
/*     */   TextField Cout;
/*     */   TextField kout;
/*     */   TextField ubout;
/*     */   boolean period = true;
/*     */ }

File: WaveTheory.js
// WaveTheory.js

window.onload = function() {
    // Create the calculator UI
    createCalculator();
};

function createCalculator() {
    let container = document.getElementById('wave-calculator');

    let form = document.createElement('div');
    form.style.display = 'grid';
    form.style.gridTemplateColumns = '200px 200px';
    form.style.gridRowGap = '10px';
    form.style.gridColumnGap = '10px';

    // Deep Water Values Title
    let title = document.createElement('h3');
    title.innerText = 'Deep Water Values:';
    title.style.gridColumn = '1 / span 2';
    form.appendChild(title);

    // Wave Height (m)
    form.appendChild(createLabel('Wave Height (m)?'));
    let Hin = createInput('1.0');
    form.appendChild(Hin);

    // Period or Frequency Choice
    let torf = createSelect(['Period', 'Frequency (Hz)']);
    form.appendChild(torf);
    let Tin = createInput('12.0');
    form.appendChild(Tin);

    // Wave Angle (degrees)
    form.appendChild(createLabel('Wave Angle (&deg;)?'));
    let angle0 = createInput('0.0');
    form.appendChild(angle0);

    // Local Depth
    form.appendChild(createLabel('Local Depth?'));
    let din = createInput('5.0');
    form.appendChild(din);

    // Buttons
    let calculateButton = document.createElement('button');
    calculateButton.type = 'button';
    calculateButton.innerText = 'Calculate';
    form.appendChild(calculateButton);

    let resetButton = document.createElement('button');
    resetButton.type = 'button';
    resetButton.innerText = 'Reset';
    form.appendChild(resetButton);

    // Placeholder elements to align the grid
    form.appendChild(document.createElement('div'));
    form.appendChild(document.createElement('div'));

    container.appendChild(form);

    // Results Section
    let results = document.createElement('div');
    results.style.display = 'grid';
    results.style.gridTemplateColumns = '200px 200px';
    results.style.gridRowGap = '10px';
    results.style.gridColumnGap = '10px';
    results.style.marginTop = '20px';

    let resultFields = {};

    let resultLabels = [
        { label: 'L (m) =', key: 'lout' },
        { label: 'k = 2&pi;/L =', key: 'kout' },
        { label: 'C = L/T =', key: 'Cout' },
        { label: 'Cg =', key: 'Cgout' },
        { label: 'n = Cg/C =', key: 'nout' },
        { label: 'Ks =', key: 'Ksout' },
        { label: 'Kr =', key: 'Krout' },
        { label: 'Angle =', key: 'angleout' },
        { label: 'H =', key: 'Hshal' },
        { label: 'u_b =', key: 'ubout' },
    ];

    resultLabels.forEach(item => {
        let label = createLabel(item.label);
        let output = createOutput();
        results.appendChild(label);
        results.appendChild(output);
        resultFields[item.key] = output;
    });

    container.appendChild(results);

    // Event Listeners
    calculateButton.addEventListener('click', function() {
        calculateWaveParameters({
            Hin: Hin.value,
            Tin: Tin.value,
            din: din.value,
            angle0: angle0.value,
            period: torf.value === 'Period',
        }, resultFields);
    });

    resetButton.addEventListener('click', function() {
        Hin.value = '';
        Tin.value = '';
        din.value = '';
        angle0.value = '0.0';
        torf.value = 'Period';
        for (let key in resultFields) {
            resultFields[key].value = '';
        }
    });
}

function createLabel(text) {
    let label = document.createElement('label');
    label.innerHTML = text;
    return label;
}

function createInput(defaultValue = '') {
    let input = document.createElement('input');
    input.type = 'text';
    input.value = defaultValue;
    return input;
}

function createSelect(options) {
    let select = document.createElement('select');
    options.forEach(opt => {
        let option = document.createElement('option');
        option.value = opt;
        option.text = opt;
        select.add(option);
    });
    return select;
}

function createOutput() {
    let output = document.createElement('input');
    output.type = 'text';
    output.readOnly = true;
    return output;
}

function calculateWaveParameters(inputs, outputs) {
    try {
        let H0 = parseFloat(inputs.Hin);
        if (isNaN(H0)) throw 'Invalid Wave Height';

        let T;
        if (inputs.period) {
            T = parseFloat(inputs.Tin);
        } else {
            let freq = parseFloat(inputs.Tin);
            if (isNaN(freq) || freq === 0) throw 'Invalid Frequency';
            T = 1 / freq;
        }
        if (isNaN(T) || T === 0) throw 'Invalid Period';

        let d = parseFloat(inputs.din);
        if (isNaN(d)) throw 'Invalid Local Depth';

        let thet0 = parseFloat(inputs.angle0);
        if (isNaN(thet0)) thet0 = 0.0;

        let refract = new Refract();

        let k = refract.waveNumber(d, T);
        outputs['kout'].value = k.toFixed(6);

        let L = 2 * Math.PI / k;
        outputs['lout'].value = L.toFixed(6);

        let C = L / T;
        outputs['Cout'].value = C.toFixed(6);

        let Angle = refract.theta(thet0, k, T);
        outputs['angleout'].value = Angle.toFixed(6);

        let Cg = refract.groupVelocity(T, d, k);
        outputs['Cgout'].value = Cg.toFixed(6);

        let n = Cg / C;
        outputs['nout'].value = n.toFixed(6);

        let Ks = refract.shoalingCoef(T, Cg);
        outputs['Ksout'].value = Ks.toFixed(6);

        let Kr = refract.refractionCoef(thet0, Angle);
        outputs['Krout'].value = Kr.toFixed(6);

        let H1 = Ks * Kr * H0;
        if (H1 > 0.8 * d) {
            H1 = 0.8 * d;
            outputs['Hshal'].value = H1.toFixed(6) + ', breaking';
        } else {
            outputs['Hshal'].value = H1.toFixed(6);
        }

        let ub = (2 * Math.PI / T) * H1 / (2 * Hyper.sinh(k * d));
        outputs['ubout'].value = ub.toFixed(6);

    } catch (error) {
        alert('Error: ' + error);
    }
}
